File: backend/src/main/java/com/sudimango/MyStudyPal/dto/AuthDto.java
package com.sudimango.MyStudyPal.dto;

import java.time.Instant;

import com.sudimango.MyStudyPal.entity.AuthProvider;
import com.sudimango.MyStudyPal.entity.User;

import jakarta.validation.constraints.NotBlank;

// @formatter:off
public class AuthDto {
    /**
     * Request
     */

    public record LoginRequest(
        @NotBlank String username,
        @NotBlank String password
    ) {}

    public record SignupRequest(
        @NotBlank String username,
        @NotBlank String password
    ) {}

    public record VerifyAccountRequest(
        @NotBlank String username,
        @NotBlank String verificationCode
    ) {}

    public record ResendVerificationEmailRequest(
        @NotBlank String username
    ) {}

    /**
     * Response
     */

    public record LoginResponse(
        String accessToken
    ) {}

    public record UserDetailsResponse(
        String userId,
        String username,
        AuthProvider authProvider,
        boolean isEnabled,
        Instant createdAt,
        Instant updatedAt
    ) {
        public UserDetailsResponse(User user) {
            this(
                user.getUserId(),
                user.getUsername(),
                user.getAuthProvider(),
                user.isEnabled(),
                user.getCreatedAt(),
                user.getUpdatedAt()
            );
        }
    }
}

File: frontend/src/lib/dto/auth-dto.ts
export enum AuthProvider {
    EMAILPASSWORD = "EMAILPASSWORD",
    GOOGLE = "GOOGLE",
}

/**
 * Request
 */

export interface LoginRequest {
    username: string;
    password: string;
}

export interface SignupRequest {
    username: string;
    password: string;
}

export interface VerifyAccountRequest {
    username: string;
    verificationCode: string;
}

export interface ResendVerificationEmailRequest {
    username: string;
}

/**
 * Response
 */

export interface LoginResponse {
    accessToken: string;
}

export interface UserDetailsResponse {
    userId: string;
    username: string;
    authProvider: AuthProvider;
    isEnabled: boolean;
    createdAt: string;
    updatedAt: string;
}
